toolchain/semantics/semantics_handle_array.cpp

```diff
@@ -2,23 +2,55 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include "toolchain/parser/parse_node_kind.h"
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleArrayExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpressionStart");
+auto SemanticsHandleArrayExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  return true;
 }
 
 auto SemanticsHandleArrayExpressionSemi(SemanticsContext& context,
                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpressionSemi");
+  context.node_stack().Push(parse_node);
+  return true;
 }
 
 auto SemanticsHandleArrayExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpression");
+  // TODO: Handle array type with undefined bound.
+  if (context.parse_tree().node_kind(context.node_stack().PeekParseNode()) ==
+      ParseNodeKind::ArrayExpressionSemi) {
+    context.node_stack().PopAndIgnore();
+    context.node_stack().PopAndIgnore();
+    return context.TODO(parse_node, "HandleArrayExpressionWithoutBounds");
+  }
+
+  auto bound_node_id = context.node_stack().PopExpression();
+  context.node_stack().PopAndIgnore();
```

Comment:
```suggestion
  context.node_stack().PopAndDiscardSoloParseNode<ParseNodeKind::ArrayExpressionSemi>();
```
We know what parse node kind we expect here, so it's better to explicitly name it. (This way we'll get an assertion failure if the expected node isn't at the top of the stack.)